examples/threading/threading.c

```diff
@@ -14,6 +24,79 @@ void* threadfunc(void* thread_param)
     // TODO: wait, obtain mutex, wait, release mutex as described by thread_data structure
     // hint: use a cast like the one below to obtain thread arguments from your parameter
     //struct thread_data* thread_func_args = (struct thread_data *) thread_param;
+  
+
+    DEBUG_LOG("Starting the thread\n");
+    struct thread_data *thread_args = (struct thread_data *) thread_param;
+
+
+    long wait_to_obtain_ns = thread_args->wait_to_obtain_ms * 1000000;
+    long wait_to_release_ns = thread_args->wait_to_release_ms * 1000000;
+
+    const struct timespec obtain_nanotime = {0, wait_to_obtain_ns};
+    const struct timespec release_nanotime = {0, wait_to_release_ns};
+
+    struct timespec remain_nanotime = {0,0};
+
+    remain_nanotime.tv_sec = 0;
+    remain_nanotime.tv_nsec = 0;
+
+    thread_args->tid = syscall(SYS_gettid);
+    
+    //usleep(thread_args->wait_to_obtain_ms * 1000);
+
+    int status = clock_nanosleep(CLOCK_REALTIME, 0, &obtain_nanotime, &remain_nanotime);
+
+    if(status != 0)
+    {
+        if(status == EINTR)
+        {
+            status = clock_nanosleep(CLOCK_REALTIME, 0, &remain_nanotime, &remain_nanotime);
+            if(status != 0)
+            {
+                return thread_param;
+            }
+        }
+        else
+        {
+            return thread_param;
+        }
+
+    }
+
+    remain_nanotime.tv_sec = 0;
+    remain_nanotime.tv_nsec = 0;    
+    
+    pthread_mutex_lock(thread_args->mutex);
+
+   // usleep(thread_args->wait_to_release_ms * 1000);
+
+
+    status = clock_nanosleep(CLOCK_REALTIME, 0, &release_nanotime, &remain_nanotime);
+
+    if(status != 0)
+    {
+        if(status == EINTR)
+        {
+            status = clock_nanosleep(CLOCK_REALTIME, 0, &remain_nanotime, &remain_nanotime);
+            if(status != 0)
+            {
+                return thread_param;
+            }
+        }
+        else
+        {
+            return thread_param;
```

Comment:
@bharathvarmaCU18 what happens to `thread_args->mutex` in these return error paths?